projects/ADuCM3029_demo_cn0428_cn0429/src/adi_m355_gas_sensor.cpp

```diff
@@ -0,0 +1,843 @@
+
+	if ((eI2cResult = adi_i2c_Close(m_i2c_handle)) == ADI_I2C_SUCCESS)
+		return SENSOR_ERROR_NONE;
+	else
+		return SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult);
+}
+
+/**
+ * @brief  Returns the temperature value
+ *
+ * @return SENSOR_RESULT
+ *
+ * @details
+ */
+SENSOR_RESULT M355_GAS::InitI2C()
+{
+	ADI_I2C_RESULT eI2cResult = ADI_I2C_SUCCESS;
+
+	eI2cResult = adi_i2c_Open(ADI_CFG_I2C_DEV_NUM, &m_I2C_memory,
+				  ADI_I2C_MEMORY_SIZE, &m_i2c_handle);
+	if (eI2cResult != ADI_I2C_SUCCESS)
+		return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+
+	eI2cResult = adi_i2c_Reset(m_i2c_handle);
+	if (eI2cResult != ADI_I2C_SUCCESS)
+		return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+
+	eI2cResult = adi_i2c_SetBitRate(m_i2c_handle, ADI_CFG_I2C_BIT_RATE);
+	if (eI2cResult != ADI_I2C_SUCCESS)
+		return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+
+	return (SENSOR_ERROR_NONE);
+}
+
+/**
+ * @brief  reads or writes data to/from the gas sensor
+ *
+ * @param  uint8_t RW, uint8_t RegAddr, uint8_t *pData, uint16_t size
+ *
+ * @return returns result of I2C communication
+ *
+ * @details
+ */
+SENSOR_RESULT M355_GAS::I2CReadWrite(uint8_t RW, uint8_t RegAddr,
+				     uint8_t *pData, uint16_t size)
+{
+	ADI_I2C_TRANSACTION sTransfer;
+	ADI_I2C_RESULT eI2cResult = ADI_I2C_SUCCESS;
+	uint32_t nHwErrors;
+	uint8_t aPrologueData[5u];
+
+	/* Set slave address for ADI M355 gas sensor */
+	eI2cResult = adi_i2c_SetSlaveAddress(m_i2c_handle, this->m_i2c_address);
+
+	if (eI2cResult != ADI_I2C_SUCCESS)
+		return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+
+	if (RW == READ) {
+		aPrologueData[0] = RegAddr;
+		sTransfer.pPrologue = &aPrologueData[0u];
+		sTransfer.nPrologueSize = 1u;
+		sTransfer.pData = pData;
+		sTransfer.nDataSize = size;
+		sTransfer.bReadNotWrite = true;
+		sTransfer.bRepeatStart = true;
+
+		eI2cResult = adi_i2c_ReadWrite(m_i2c_handle, &sTransfer,
+					       &nHwErrors);
+		if (eI2cResult == ADI_I2C_SUCCESS) {
+			return (SENSOR_ERROR_NONE);
+		}
+	}
+
+	else if (RW == WRITE) {
+		aPrologueData[0] = RegAddr;
+		sTransfer.pPrologue = &aPrologueData[0u];
+		sTransfer.nPrologueSize = 1u;
+		sTransfer.pData = pData;
+		sTransfer.nDataSize = size;
+		sTransfer.bReadNotWrite = false;
+		sTransfer.bRepeatStart = true;
+
+		eI2cResult = adi_i2c_ReadWrite(m_i2c_handle, &sTransfer,
+					       &nHwErrors);
+		if (eI2cResult == ADI_I2C_SUCCESS) {
+			return (SENSOR_ERROR_NONE);
+		}
+	}
+
+	return (SET_SENSOR_ERROR(SENSOR_ERROR_I2C, eI2cResult));
+}
+
+/**
+ * @brief  initializes the sensor - requesting sensor status
+ * 			until 0x01 response received (IDLE)
+ *
+ * @param  uint8_t sensor_address
+ *
+ * @return returns result of I2C communication
+ *
+ * @details
+ */
+SENSOR_RESULT M355_GAS::SensorInit(uint8_t sensor_address)
+{
+	uint8_t pBuff[2] = { 0 };
+	uint16_t timeout = 5;
+	SENSOR_RESULT Result = this->openWithAddr(sensor_address);
+
+	do {
+		Result = this->I2CReadWrite(READ, GET_STATUS, &pBuff[0], 1u);
```

Comment:
Because this is an initialization function, I think it was intended to give the sensor 500ms to respond in case it is not ready the first time. If the slave is still in the process of initializing its I2C communications, then the I2CReadWrite would fail, but we still want to give the slave a chance to respond the next time through the loop, which won't happen if we return. Would you suggest removing the loop altogether?

Reply:
No, in this case I think the loop makes sense. Probably there is no point in writing the Result variable here and maybe that should be removed.